File: internal/manifest/manifest.go
//go:build goexperiment.jsonv2

package manifest

import (
	"encoding/json/v2"
	"encoding/json/jsontext"
)

// Manifest represents the structure of the manifest configuration file
// according to the JSON schema specification in plan.md section 10
type Manifest struct {
	Project  ProjectConfig  `json:"project"`
	Scan     ScanConfig     `json:"scan"`
	Limits   *LimitsConfig  `json:"limits,omitempty"`
	Cache    *CacheConfig   `json:"cache,omitempty"`
	Features *FeatureConfig `json:"features,omitempty"`
}

// ProjectConfig contains project-specific configuration
type ProjectConfig struct {
	Root     string `json:"root"`
	Composer string `json:"composer"`
}

// ScanConfig contains scanning configuration settings
type ScanConfig struct {
	Targets         []string `json:"targets"`
	VendorWhitelist []string `json:"vendor_whitelist,omitempty"`
	Globs           []string `json:"globs,omitempty"`
}

// LimitsConfig contains analysis limits and constraints
type LimitsConfig struct {
	MaxWorkers *int `json:"max_workers,omitempty"`
	MaxFiles   *int `json:"max_files,omitempty"`
	MaxDepth   *int `json:"max_depth,omitempty"`
}

// CacheConfig contains caching configuration
type CacheConfig struct {
	Enabled *bool   `json:"enabled,omitempty"`
	Kind    *string `json:"kind,omitempty"`
}

// FeatureConfig controls which analysis features are enabled
type FeatureConfig struct {
	HTTPStatus        *bool `json:"http_status,omitempty"`
	RequestUsage      *bool `json:"request_usage,omitempty"`
	ResourceUsage     *bool `json:"resource_usage,omitempty"`
	WithPivot         *bool `json:"with_pivot,omitempty"`
	AttributeMake     *bool `json:"attribute_make,omitempty"`
	ScopesUsed        *bool `json:"scopes_used,omitempty"`
	Polymorphic       *bool `json:"polymorphic,omitempty"`
	BroadcastChannels *bool `json:"broadcast_channels,omitempty"`
}

// ToJSON serializes the manifest to JSON bytes.
func (m *Manifest) ToJSON() ([]byte, error) {
	return json.Marshal(m, json.Deterministic(true), jsontext.WithIndent("  "))
}

// Backward compatibility type alias for legacy code
type Features = FeatureConfig

File: src/manifest.rs
use std::path::{Path, PathBuf};

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Manifest {
    pub project: ProjectConfig,
    #[serde(default)]
    pub scan: ScanConfig,
    #[serde(default)]
    pub limits: LimitsConfig,
    #[serde(default)]
    pub cache: CacheConfig,
    #[serde(default)]
    pub features: FeatureFlags,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProjectConfig {
    pub root: PathBuf,
    #[serde(default)]
    pub composer: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScanConfig {
    #[serde(default = "default_targets")]
    pub targets: Vec<String>,
    #[serde(default = "default_globs")]
    pub globs: Vec<String>,
    #[serde(default)]
    pub vendor_whitelist: Vec<String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct LimitsConfig {
    #[serde(default)]
    pub max_workers: Option<usize>,
    #[serde(default)]
    pub max_files: Option<usize>,
    #[serde(default)]
    pub max_depth: Option<usize>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct CacheConfig {
    #[serde(default)]
    pub enabled: bool,
    #[serde(default)]
    pub kind: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct FeatureFlags {
    #[serde(default = "default_true")]
    pub http_status: bool,
    #[serde(default = "default_true")]
    pub request_usage: bool,
    #[serde(default = "default_true")]
    pub resource_usage: bool,
    #[serde(default = "default_true")]
    pub with_pivot: bool,
    #[serde(default = "default_true")]
    pub attribute_make: bool,
    #[serde(default = "default_true")]
    pub scopes_used: bool,
    #[serde(default = "default_true")]
    pub polymorphic: bool,
    #[serde(default = "default_true")]
    pub broadcast_channels: bool,
}

impl Default for ScanConfig {
    fn default() -> Self {
        Self {
            targets: default_targets(),
            globs: default_globs(),
            vendor_whitelist: Vec::new(),
        }
    }
}

impl Default for FeatureFlags {
    fn default() -> Self {
        Self {
            http_status: true,
            request_usage: true,
            resource_usage: true,
            with_pivot: true,
            attribute_make: true,
            scopes_used: true,
            polymorphic: true,
            broadcast_channels: true,
        }
    }
}

impl Manifest {
    pub fn resolve_paths(&mut self, manifest_path: &Path) {
        let base_dir = manifest_path.parent().unwrap_or_else(|| Path::new("."));
        self.project.root = resolve_path(base_dir, &self.project.root);
    }
}

fn default_targets() -> Vec<String> {
    vec!["app".to_string(), "routes".to_string()]
}

fn default_globs() -> Vec<String> {
    vec!["**/*.php".to_string()]
}

fn default_true() -> bool {
    true
}

fn resolve_path(base_dir: &Path, path: &Path) -> PathBuf {
    if path.is_absolute() {
        return path.to_path_buf();
    }

    if path.exists() {
        return path.to_path_buf();
    }

    let manifest_relative = base_dir.join(path);
    if manifest_relative.exists() {
        return manifest_relative;
    }

    manifest_relative
}
